### PR TITLE
chore: Add FLUX layers

### LINKED ISSUE #95: [OTHER] Add Flux layers
## Summary

Implement 'layers.py' file for Flux model.

### PR DESCRIPTION
**Pull request by aturker-synnada**
## Description

Closes #95 

## What is Changed

Include the changes introduced in this PR.

- FLUX layers added

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: examples/flux/layers.py [hunk 1]
```diff
@@ -13,17 +13,40 @@
...
+    ScaledDotProduct,
+    Sigmoid,
+    SiLU,
+    Split,
+    Transpose,
+)
 
 
-def rms_norm(dim: int) -> Model:
+def mlp_embedder(hidden_dim, name: str | None = None):
```
Comment (kberat-synnada) [inline examples/flux/layers.py:33]: fill all missing type hints 

e.g. def mlp_embedder(hidden_dim: int, name: str | None = None) -> Model:

File: examples/flux/layers.py [hunk 2]
```diff
@@ -75,4 +97,233 @@ def attention() -> Model:
...
+
+    return block
+
+
+def qk_norm(dim: int, name: str | None = None):
+    block = Model(name=name)
+    query_norm = rms_norm(dim, name="query_norm")
+    key_norm = rms_norm(dim, name="key_norm")
+
+    query_norm.name = "query_norm"
```
Comment (kberat-synnada) [inline examples/flux/layers.py:109]: add name argument to rms_norm function

Comment (mehmetozsoy-synnada) [inline examples/flux/layers.py:109]: setting name seems unnecessary as its name is already set inside `rms_norm`

File: examples/flux/layers.py [hunk 3]
```diff
@@ -75,4 +97,233 @@ def attention() -> Model:
...
+    img_attn = block.attn[:, 256:]  # type: ignore[attr-defined]
+
+    block += Linear(hidden_size, name="img_attn_proj")(img_attn, output="img_proj")
+    img = img + block.img_mod_1[2] * block.img_proj  # type: ignore[attr-defined]
+
+    block += LayerNorm(use_scale=False, use_bias=False, name="img_norm2", eps=1e-6)(
+        img, output="img_norm_2"
+    )
+    img_norm_2 = block.img_norm_2  # type: ignore[attr-defined]
+
```
Comment (kberat-synnada) [inline examples/flux/layers.py:217]: we could encapsulate this part into a function for img_mlp and txt_mlp:

    img_mlp = Model(name="img_mlp")
    img_mlp += Linear(mlp_hidden_dim, name="0")(input="input")
    img_mlp += Gelu(approximate=True)
    img_mlp += Linear(hidden_size, name="2")(output="output")

Comment (kberat-synnada) [reply examples/flux/layers.py:217]: mlp_embedder could be used 

e.g. def mlp_embedder(hidden_dim, activation: Model, name: str | None = None):

File: examples/flux/layers.py [hunk 4]
```diff
@@ -75,4 +97,233 @@ def attention() -> Model:
...
+    block += LayerNorm(use_scale=False, use_bias=False, eps=1e-6, name="img_norm1")(
+        input=img, output="img_norm"
+    )
+
+    img_modulated = (1 + block.img_mod_1[1]) * block.img_norm + block.img_mod_1[0]  # type: ignore[attr-defined]
+
+    block += Linear(hidden_size * 3, use_bias=qkv_bias, name="img_attn_qkv")(
+        img_modulated, output="img_qkv"
+    )
+
```
Comment (kberat-synnada) [inline examples/flux/layers.py:158]: Could we make all rearrange operations (for img, txt and single stream) in one function like following:
    
    def add_rearrange_model(block, qkv_connection, hidden_size, num_heads, key_identifier = "")
        img_qkv_shape = qkv_connection.shape()  # type: ignore[attr-defined]
        B, L = img_qkv_shape[0], img_qkv_shape[1]
        block += Reshape()(shape=(B, L, 3, num_heads, -1))
        block += Transpose(axes=(2, 0, 3, 1, 4))(output=key_identifier+"transpose_out")
        img_q, img_k, img_v = (
            block.transpose_out[0],  # type: ignore[attr-defined]
            block.transpose_out[1],  # type: ignore[attr-defined]
            block.transpose_out[2],  # type: ignore[attr-defined]
        )
        block += qk_norm(hidden_size // num_heads, name=key_identifier+"attn_norm")(
            q_in=img_q, k_in=img_k, q_out=key_identifier+"q_out", k_out=key_identifier+"k_out"
        )
        img_q, img_k = block.q_out, block.k_out  # type: ignore[attr-defined]
        return img_q, img_k, img_v

File: examples/flux/layers.py [hunk 5]
```diff
@@ -75,4 +97,233 @@ def attention() -> Model:
...
+        block += Buffer()(modulation[1].split(3, axis=-1), IOKey("mod_2"))
+    else:
+        modulation = lin_out.split(3, axis=-1)
+        block += Buffer()(modulation, IOKey("mod_1"))
+
+    return block
+
+
+def double_stream_block(
+    hidden_size: int, num_heads: int, mlp_ratio: float, qkv_bias: bool = False
```
Comment (mehmetozsoy-synnada) [inline examples/flux/layers.py:136]: in this block. Overall, It seems like operations done for `img` and operations done in `txt` seems very similar. Can these be grouped?

File: examples/flux/layers.py (hunk 3 above)
Comment (aturker-synnada) [reply examples/flux/layers.py:217]: Cannot create them in same function, naming of primitives are different

Comment (aturker-synnada) [reply examples/flux/layers.py:217]: Instead of creating a new function, what about using deepcopy?

**Review by kberat-synnada — APPROVED** [review]

**Review by norhan-synnada — APPROVED** [review]

**Pull request merged**